microdnf/plugins.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright Contributors to the libdnf project.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "plugins.hpp"
+
+#include "library.hpp"
+#include "utils/bgettext/bgettext-lib.h"
+
+#include <fmt/format.h>
+
+#include <filesystem>
+
+namespace microdnf {
+
+// Support for Plugin in the shared library.
+class PluginLibrary : public Plugin {
+public:
+    // Loads a shared library, finds symbols, and instantiates the plugin.
+    explicit PluginLibrary(const std::string & library_path);
+
+    ~PluginLibrary();
+
+private:
+    using TGetApiVersionFunc = decltype(&microdnf_plugin_get_api_version);
+    using TGetNameFunc = decltype(&microdnf_plugin_get_name);
+    using TGetVersionFunc = decltype(&microdnf_plugin_get_version);
+    using TNewInstanceFunc = decltype(&microdnf_plugin_new_instance);
+    using TDeleteInstanceFunc = decltype(&microdnf_plugin_delete_instance);
+    TGetApiVersionFunc get_api_version{nullptr};
+    TGetNameFunc get_name{nullptr};
+    TGetVersionFunc get_version{nullptr};
+    TNewInstanceFunc new_instance{nullptr};
+    TDeleteInstanceFunc delete_instance{nullptr};
+    utils::Library library;
+};
+
+PluginLibrary::PluginLibrary(const std::string & library_path) : library(library_path) {
+    get_api_version = reinterpret_cast<TGetApiVersionFunc>(library.get_address("microdnf_plugin_get_api_version"));
+    get_name = reinterpret_cast<TGetNameFunc>(library.get_address("microdnf_plugin_get_name"));
+
+    auto api_version = get_api_version();
+    if (api_version.major != PLUGIN_API_VERSION.major || api_version.minor > PLUGIN_API_VERSION.minor) {
+        auto msg = fmt::format(
+            "Unsupported plugin API combination. API version required by \"{}\" plugin (\"{}\") is \"{}.{}\"."
```

Comment:
Sorry about being nitpicky, can you put the name after the word "plugin"? This way it looks weird:
> Unsupported plugin API combination. API version required by plugin \"{}\" (\"{}\") is ...